Shipping Labels: Add create custom package form when there are existing active packages

hafizrahman opened the pull request:
# What this PR is for:

This PR adds a functionality to allow users to add a custom package, during shipping label creation process. 

There are several cases where the "add custom package" form can be shown. This PR specifically handles the case where the "add custom package" functionality is added where the user's site already has one or more packages activated already.

# Testing instructions:

## Testing phase 1: Going to the "Add new package" screen:

0. Make sure the site has Shipping Labels creation enabled, and have an Order where Shipping Labels can be created.
1. In the app, find that Order and create a Shipping Label,
2. Complete the "Ship from" and "Ship to" steps,
3. Enter "Packaging details" step,
4. Tap the "Package selected" spinner, this will open the "Selected Package" screen where you can select an existing package to use,
5. At the bottom of the screen, make sure the "Create New Package" button is shown,
6. Tap the button,
7. The "Add new package" screen will be shown.

## Testing phase 2.A: Creating a new custom package, successful path
1. On the "Add new package" screen, make sure "Custom Package" tab is selected,
2. Fill in all the fields, especially take note of the entered name and weight values, then tap Done (on top right),
3. Make sure the loading dialog is shown while the API call is being done,
4. Once it is completed, make sure the screen is automatically moved back to the "Package details" screen,
5. Make sure a snackbar is shown at the bottom saying that "[package name] saved",
6. Make sure the name of the recently created package is shown in the "Package selected" field,
7. Make sure the weight of the recently created package is shown in the "Total package weight" field
8. Tap the "Package selected" spinner,
9. On the "Selected package" screen, make sure the newly created package is listed in the "Custom packages" list.

## Testing phase 2.B: Creating a new custom package, error path
1. On the "Add new package" screen, make sure "Custom Package" tab is selected,
2. Tap Done (on top right) immediately. Make sure that all fields starting from "Package name" are showing error messages, "This field is required",
3. Also make sure a snackbar saying "Package creation failed. Please try again" is shown at the bottom,
4. Enter a zero value in the "Length" field. Make sure the error message "Invalid value" is shown,
5. Fill in everything, and in the Package Name enter a name that is identical with an existing package (for example the Package that was created on phase 2.A). Make sure a snackbar is shown, saying "Package creation failed: At least one of the new custom packages has the same name as existing packages".

# Known limitations:
- No unit tests yet. This will be added once all the other "add custom package" cases are added.
- So far I've been incorrectly using `./gradlew detekt` instead of `./gradlew detektall`, and just noticed how CircleCi is failing and that there's a lot of small fixes it recommends. I'll work on that ASAP but for now I think the PR can be reviewed regardless of that.

# Screenshots

| Screenshot | Description |
| - | - |
| <img src="https://user-images.githubusercontent.com/266376/124588414-f6a42780-de82-11eb-8914-b64d0b447dcc.png" width="420px" /> | "Create New Package" button is added at the bottom of the package selector screen |
| <img src="https://user-images.githubusercontent.com/266376/124588653-3ff47700-de83-11eb-80da-9347cd7faa35.png" width="420px" /> | Custom package creation form |
| <img src="https://user-images.githubusercontent.com/266376/124588787-66b2ad80-de83-11eb-97fe-b7954eafe244.png" width="420px" /> | Form showing error states |
| <img src="https://user-images.githubusercontent.com/266376/124588885-88139980-de83-11eb-9cbd-2310dea644c7.png" width="420px" /> | After successful package creation: Note that package name "2a" is shown on Snackbar and on the "Package selected" spinner |





Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

hafizrahman committed:
Add "Create new package" button to package selector screen.

hafizrahman committed:
Add click listener to create package button and create event for it.

hafizrahman committed:
Start tabs architecture using ViewPager2.

Tabs are initialized but actual pager adapter functionalities (for showing / hiding the forms) are not completed yet.

hafizrahman committed:
Continue ViewPager2 setup, added view model as well.

This also adds the custom and service package fragments to be shown in the pager, although it's just the files and they're now still empty.

hafizrahman committed:
Add the create package screen to the navigation component.

The "add new package" button will now open properly to the create package fragment.

hafizrahman committed:
Set proper title for the create package screen.

hafizrahman committed:
Add custom package form creation layout.

hafizrahman committed:
Setup create custom package form's package type selection input.

The selection dialog does not yet open on tap, though. Still trying to figure it out.

hafizrahman committed:
Simplify spinner to use the setup() method.

This avoids the need to create a custom DialogFragment and use navigation component.

hafizrahman committed:
Add menu item "Done" for create custom package form.

The item doesn't process the data yet.

hafizrahman committed:
Start work for real-time form input validation.

This now already works for the "Length" input field, but not yet added to the other fields.

Also adds some initial work for data submission (marked with "TODO: for data submission"), but they're incomplete yet.

hafizrahman committed:
Add real-time validation for the rest of custom package input fields.

hafizrahman committed:
Keyboard form scrolling fix.

hafizrahman committed:
Merge branch 'develop' into add/create-custom-package-on-existing-packages-case

hichamboushaba committed:
Fix resize issue

hafizrahman committed:
Merge branch 'hicham/fix-activity-resize' into add/create-custom-package-on-existing-packages-case

# Conflicts:
#	WooCommerce/src/main/res/layout/fragment_shipping_label_create_package.xml

hafizrahman committed:
Adds validation process when "Done" menu is tapped.

Also adds general refactoring to the entire validation process.

hafizrahman committed:
Implement D.I. and move validation to viewmodel layer.

hafizrahman committed:
Implement D.I. on ShippingLabelCreatePackageFragment as well.

hafizrahman committed:
Add API call to create the package.

This works, but success/error message handling in the app still has to be added.

hafizrahman committed:
Refactor viewmodel usage so that each form now has its own.

hafizrahman committed:
Try a mechanism to go back from adding package to package edit screen.

This commit will likely crash the app because it's using hardcoded value in ShippingLabelCreatePackageViewModel.

hafizrahman committed:
Pass `position` value to be re-used.

We're passing it from `EditShippingLabelPackagesFragment` to `ShippingPackageSelectorFragment` to `ShippingLabelCreatePackageViewModel`, so that after a custom package is created, the position value can be used in the the `navigateBackWithResult()` call in `ShippingLabelCreatePackageFragment`

hafizrahman committed:
Add success and error handling after DONE button is tapped.

Success: return user to the EditPackage screen, make sure the created package is selected, and Snackbar is shown informing user that the new package is saved.

Failure: Show Snackbar with error if there's API or unknown error after submitting.

hafizrahman committed:
Add saving progress dialog.

hafizrahman committed:
Merge branch 'develop' into add/create-custom-package-on-existing-packages-case

hafizrahman committed:
Update error messages to match core; also refactor validation in general.

hafizrahman committed:
Add proper units to the input fields that use them.

hafizrahman committed:
UI tweak: hide weight input helper text if error is shown.

hafizrahman committed:
Various code cleanups and comment addition as needed.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt:
```diff
@@ -17,14 +19,27 @@ data class ShippingPackage(
         const val CUSTOM_PACKAGE_CATEGORY = "custom"
     }
+
+    fun toCustomPackageAppModel(): CustomPackage {
```
np, I would rename this to `toCustomPackageDataModel` as the `CustomPackage` is in the data layer, and not business layer.

hafizrahman replied:
Thanks, good catch!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/model/ShippingPackage.kt:
```diff
@@ -17,14 +19,27 @@ data class ShippingPackage(
     val height: Float
-) : Parcelable
+) : Parcelable {
+    fun toDimensionsString() : String {
```
np, I would just `toString` instead of a new function `toDimensionsString`, this will follow the conventions better, and will make sure that the correct output is sent when this is concatenated to a string...

hafizrahman replied:
Ah yes. Initially I wanted to use `toString()` , then I saw that it had to be an override and I wasn't sure if that's safe. But now that I think of it, I don't think we'll need to do any different version of string conversion soon anyway. Thanks for explaining.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/ShippingLabelRepository.kt:
```diff
@@ -205,6 +205,26 @@ class ShippingLabelRepository @Inject constructor(
+        ).let { result ->
+            when {
+                result.model == true -> {
+                    // Add package to existing list of available custom packages, so that it will
+                    // be available immediately in ShippingPackageSelectorFragment if user visits that page
+                    // again after creating a custom package.
```
Personally, I would remove this comment, IMO it's stating something obvious, and also it's coupling it to the `ShippingPackageSelectorFragment`, which is not the only reason, as the real reason is to avoid having outdated data in the repository.
Comments shouldn't be added unless the code is not clear and obvious by itself, because they add to the noise, and require their own maintenance, for example, if you refactor the code and move something to another place, you would have to think about the comment, and whether it's still valid, check this discussion around the topic: https://softwareengineering.stackexchange.com/a/176446

hafizrahman replied:
Thanks for explaining and including a link too!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/EditShippingLabelPackagesViewModel.kt:
```diff
@@ -157,6 +157,16 @@ class EditShippingLabelPackagesViewModel @Inject constructor(
         }
         viewState = viewState.copy(shippingLabelPackages = packages)
+
+        // When a package is successfully created inside the app, user is immediately returned from the creation form
+        // to this Edit Shipping Label screen. In this situation, we show a Snackbar to inform the user
+        // that the creation is completed successfully.
+        if(isFromPackageCreation) {
+            triggerEvent(ShowSnackbar(
+                message = string.shipping_label_create_custom_package_success_message,
+                args = arrayOf(selectedPackage.title)
+            ))
+        }
```
I would move this logic to the real concerned part of code: `ShippinLabelCreatePackageFragment`, as it's their responsibility, and not here, this would save some code by not having to think about `isFromPackageCreation`, and would be cleaner.
You can do so by calling `triggerEvent(ShowSnackBar)` in this part before triggering the exit event: https://github.com/woocommerce/woocommerce-android/blob/59ea827b79362836745fd5325a2db676695d5bc0/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageViewModel.kt#L21
And this is something we follow a lot in the app, for example [here](https://github.com/woocommerce/woocommerce-android/blob/697172b843b1357f402808f24d829fe13f91a721/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/notes/AddOrderNoteViewModel.kt#L99)

hafizrahman replied:
Oh, that's new to me. I didn't realize that a Snackbar can still show up *after* navigating away, that's why I made it appear in `EditShippingLabelPackagesFragment` initially. Thanks for pointing that out!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageFragment.kt:
```diff
@@ -0,0 +1,191 @@
+        } else {
+            inputLayout.error = resources.getString(message)
+        }
+    }
```
This is overcomplicating the logic a bit, you can just change the above observers to something like this:
```kotlin
            new.customFormNameError.takeIfNotEqualTo(old?.customFormNameError) { error ->
                binding.customPackageFormName.error = error?.let { getString(it) }
            }
```
As setting the error to `null` will automatically disable the error UI.
Regarding the height, if it's an issue, we should update our `WCMaterialOutlinedEditTextView` to fix it instead of just fixing it here, and I wouldn't worry about in this PR 🙂

hafizrahman replied:
Makes sense! I've fixed that, and added in my to-do list to create a GH issue for the WCMaterialOutlinedEditTextView behavior.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+        viewState = viewState.copy(customFormType = selectedPackageType)
+    }
+
+    fun sanitizeStringInput(input: String) {
+        if(input.isBlank()) {
+            updateErrorViewState(InputName.NAME, emptyInputError)
+        }
+        else {
+            updateErrorViewState(InputName.NAME, null)
+        }
+    }
```
There is small issue here, the name of the function suggests that it will be used for multiple input fields, but the body is hardcoding the `InputName.Name`.

Personally I would change the logic of the viewmodel, and use a single function like this:
```kotlin
fun onFieldTextChanged(value: String, field: InputName) {
      when(field) {
             InputName.Name -> validateNameField(value)
             else -> validateFloatInput(input: String, name: InputName)
      }
}

private fun validateNameField(input: String) {
    // same as your logic
}

private fun validateFloatInput(input: String, field: inputName) {
    // same as your logic
}
```

This will reduce the API interface of the ViewModel and make adding new fields easier, WDYT?

hafizrahman replied:
Thanks! That's clever and does make it more extensible.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    fun sanitizeFloatInput(input: String, name: InputName) {
+        val acc = inputToFloat(input)
```
WDYT about making `inputToFloat` as a [local function](https://kotlinlang.org/docs/functions.html#local-functions), it will reduce the number of functions of the ViewModel, and make less noise.

hafizrahman replied:
That makes sense! However in my current code, inputToFloat is used in a couple of places (`validateFloatInput()` and `onCustomFormDoneMenuClicked()`).

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+        }
+    }
+
+    fun onCustomFormDoneMenuClicked(type: String, name: String,
+                                    length: String, width: String,
+                                    height: String, weight: String) {
```
You don't need to change anything here, I just wanted to share something about a better separation of responsibilities and better data flow here.

Personally, I avoid passing the data in the `done` button listener, instead I would save all the data to the viewstate initially when validating it, and when the `done` button is clicked, I would read the data from the viewstate, and not from the fields.
We follow a similar things in multiple screens in the app, for example this simple [one](https://github.com/woocommerce/woocommerce-android/blob/697172b843b1357f402808f24d829fe13f91a721/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/downloads/ProductDownloadDetailsViewModel.kt#L52-L73).

hafizrahman replied:
I like that! That definitely feels better (and as a side-effect also helps with the Detekt issue of having too many parameters).

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+        sanitizeFloatInput(weight, InputName.EMPTY_WEIGHT)
+
+        // At this point, if there's no errors, we assume everything is valid and good to go.
+        if(viewState.areAllRequiredFieldsValid) {
```
To reduce the indent, it's recommended to return early here, as we don't need to do any operation unless the data if validated.
```kotlin
if (!viewState.areAllRequiredFieldsValid) return
// rest of code without indentation
```

hafizrahman replied:
I like the idea of returning early. Thanks!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -0,0 +1,162 @@
+            }
+        }
+        else {
+            triggerEvent(ShowSnackbar(string.shipping_label_create_custom_package_generic_failure))
```
This will trigger the generic error if the fields are not validated!! is this intended?

hafizrahman replied:
Initially I did want it to show, but now that I think of it I agree it's excessive as it's already showing error message on the input fields.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageFragment.kt:
```diff
@@ -0,0 +1,82 @@
+
+@AndroidEntryPoint
+class ShippingLabelCreatePackageFragment: BaseFragment(R.layout.fragment_shipping_label_create_package) {
+    private var _binding: FragmentShippingLabelCreatePackageBinding? = null
+    private val binding get() = _binding!!
+
+    private var _tabLayout: TabLayout? = null
+    private val tabLayout get() = _tabLayout!!
+
+    private var _viewPager: ViewPager2? = null
+    private val viewPager get() = _viewPager!!
```
Instead of saving the UI parts as fields of the fragment, and having to clean them up in `onDestroyView`, it's recommended to have them as local variables in `onViewCreated`, and pass them to functions that need them as arguments, this way the GC will clear them for us without having to nullify them.

For example:
```kotlin
override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val tabLayout = binding.createPackageTabLayout
        val viewPager = binding. createPackagePager
        initializeTabs(tabLayout, viewPager)
}

private fun initializeTabs(tabLayout: TabLayout, viewPage: ViewPager) {
        // ...
        TabLayoutMediator(tabLayout, viewPager) { tab, position ->
            //...
        }.attach()
    }
```

hafizrahman replied:
I see, that makes sense!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageFragment.kt:
```diff
@@ -0,0 +1,82 @@
+
+    private fun initializeTabs() {
+        // Get the english version to use for setting the tab tag.
+        val englishTabArray = StringUtils
+            .getStringArrayByLocale(requireContext(), R.array.shipping_label_create_new_package_tabs, "en")
+
+       val tabArray = resources.getStringArray(R.array.shipping_label_create_new_package_tabs).toList()
+        TabLayoutMediator(tabLayout, viewPager) { tab, position ->
+            tab.text = tabArray[position]
+            tab.tag = englishTabArray?.get(position) ?: tabArray[position]
+        }.attach()
```
Do we need this logic? I don't see any need for setting the `tag` here

hafizrahman replied:
I wasn't sure if the tag will be needed or not later on when working on the second tab, so I added this just in case (based on the example on `OrderListFragment`). But yeah, it's likely better to just remove it for now and re-add if needed.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageFragment.kt:
```diff
@@ -0,0 +1,82 @@
+    private fun setupObservers(viewModel: ShippingLabelCreatePackageViewModel) {
+        viewModel.event.observe(viewLifecycleOwner) { event ->
+            when (event) {
+                is SelectPackageEvent -> {
```
We don't need a new event for this 🙂, you can use `ExitWithResult` to stay consistent with the rest of the app.

hafizrahman replied:
Ah got it!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageViewModel.kt:
```diff
@@ -0,0 +1,45 @@
+
+    data class SelectPackageEvent(
+        val packageResult: ShippingPackageSelectorResult) : MultiLiveEvent.Event()
+}
```
See above about this class 🙂, it can be removed once you migrate to ExitWithResult

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageViewPagerAdapter.kt:
```diff
@@ -0,0 +1,17 @@
+import androidx.viewpager2.adapter.FragmentStateAdapter
+import com.woocommerce.android.ui.orders.shippinglabels.creation.ShippingLabelCreatePackageViewModel.PackageType
+
+class ShippingLabelCreatePackageViewPagerAdapter(container: Fragment, val itemsCount: Int)
```
I don't think we need to pass the itemsCount as parameter here?
We have two possible cases:
1. We are sure we won't have any other tabs, and then `getItemCount` should be hardcoded to 2.
2. We want to stay open to any possibility, then change `getItemCount` logic to
```kotlin
override fun getItemtCount() = PackageType.values().size
```

hafizrahman replied:
Ah true, that was unneeded. I'm going with case 2 for now, thanks!

hichamboushaba commented on WooCommerce/src/main/res/layout/fragment_shipping_label_create_custom_package.xml:
```diff
@@ -0,0 +1,114 @@
+            android:id="@+id/custom_package_form_info"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/major_100"
+            android:layout_marginTop="@dimen/major_75"
+            android:layout_marginEnd="@dimen/major_100"
```
Just a tip, but you don't need to implement it, for screens where the margin is defined for most or all views, we can use ConstraintLayout's [guidelines](https://developer.android.com/reference/androidx/constraintlayout/widget/Guideline), an example of this is available [here](https://github.com/woocommerce/woocommerce-android/blob/dde15dc823e78e11f362e03609939a5596565c12/WooCommerce/src/main/res/layout/shipping_customs_list_item.xml#L8)

hafizrahman replied:
Ah yeah! I remember you mentioning about guidelines before. Thanks for reminding me! Will play with it for the next layout I'm working on as needed.

hichamboushaba commented on WooCommerce/src/main/res/layout/fragment_shipping_label_create_custom_package.xml:
```diff
@@ -0,0 +1,114 @@
+            app:layout_constraintTop_toTopOf="parent" />
+
+        <com.woocommerce.android.widgets.WCMaterialOutlinedSpinnerView
+            android:id="@+id/custom_package_form_type"
```
A remark regarding the view ids, having the `custom_package_form` is a noise IMO, we needed it when we were kotlin synthetics, as we need a way to encapsulate and differentiate between views of each screen, but with ViewBinding, as the views are already encapsulated to the `binding` object, we don't need it, so personally, I would name this: `@+id/type_spinner`

hafizrahman replied:
Oh, thanks for this. The verbosity did bother me as well but I wasn't sure what the best practice is.

hichamboushaba commented on WooCommerce/src/main/res/layout/fragment_shipping_packages_selector.xml:
```diff
@@ -17,11 +17,55 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent">
+
+        <View
+            android:id="@+id/packages_createNewDivider"
+            style="@style/Woo.Divider"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent" />
```
I don't think we need this divider, as in Figma we have only the shadow, which will be set by the elevated layout.
And if we can remove it, then we can remove one level of nesting by moving `WCElevatedLinearLayout` to be a direct child of the main layout.

hafizrahman replied:
Oh I see! The shadow was a bit faint for me, especially in dark mode, but I agree that it does exist and matches Figma as well.

hichamboushaba reviewed (commented):
Truly great job @hafizrahman 👏 

I left some comments and tips, the main goal is to learn from this experience as much as possible.

And one point that I want to emphasize, please pay attention to the comments point I mentioned, as unneeded comments that are stating the obvious will become a big burden to maintain very fast.

hafizrahman committed:
Various small refactoring.

hafizrahman committed:
Various refactoring in viewmodel (plus the affected fragment).

Mostly about better validation handling, as well as saving input values in viewstate so that Done menu handling does not depend on getting the data again.

hafizrahman committed:
Refactor the way successful saving Snackbar is shown.

hafizrahman committed:
Various fragment/view/layout refactoring.

Also includes simplification for ExitWithResult event.

hichamboushaba commented on WooCommerce/src/main/res/layout/fragment_shipping_label_create_custom_package.xml:
```diff
@@ -96,18 +96,18 @@
 
         <com.google.android.material.textview.MaterialTextView
-            android:id="@+id/custom_package_form_empty_weight_helper"
+            android:id="@+id/weight_helper"
```
Sorry I just noticed this, and it's not a blocker if you left it like this, I just wanted to give you a tip that the material TextInputLayout accepts a `helperText` natively, so you can just add `app:helperText="@string/shipping_label_create_custom_package_field_empty_weight_helper"` to the view above without needing another view, and thinking about its constraints.

hafizrahman replied:
Oh that is a neat tip. This also automatically resolves the "hide hint if there's error" situation. Added!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreatePackageFragment.kt:
```diff
@@ -10,73 +10,63 @@ import com.woocommerce.android.R
 class ShippingLabelCreatePackageFragment: BaseFragment(R.layout.fragment_shipping_label_create_package) {
+    @Inject lateinit var uiMessageResolver: UIMessageResolver
+
     private var _binding: FragmentShippingLabelCreatePackageBinding? = null
```
My remark about making views local variables of `onViewCreated` concerns `binding` too 🙂, I'm mentioning it just for future PRs, it's not a big issue, you can leave it like this in this PR.

hafizrahman replied:
Ah, gotcha! I'll move it since I'll likely be using this code for reference in the future as well.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -42,89 +42,101 @@ class ShippingLabelCreateCustomPackageViewModel @Inject constructor(
+        updateInputInViewState(value, field)
+    }
+
+    fun validateNameField(input: String) {
```
I think this can be private.

hafizrahman replied:
Good catch!

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/shippinglabels/creation/ShippingLabelCreateCustomPackageViewModel.kt:
```diff
@@ -42,89 +42,101 @@ class ShippingLabelCreateCustomPackageViewModel @Inject constructor(
     }
 
-    fun sanitizeFloatInput(input: String, name: InputName) {
+    fun validateFloatInput(input: String, name: InputName) {
```
Same as above

hichamboushaba reviewed (commented):
Thanks @hafizrahman for making the changes, this really looks very nice.
Once you fix the detekt issues, ping me for a final review and to merge it

hafizrahman committed:
Various `detektAll`-related formatting fixes.

hafizrahman committed:
Various small refactoring again.

Most notably:
- add binding as local variable in onViewCreated
- use built in input helper instead of a separate view.
- make private some functions

hafizrahman commented:
Ready for final review @hichamboushaba . Thank you!

hichamboushaba commented:
Looks really nice.
One final remark, I think this better be merged to `develop` and not a feature branch, this would help avoid dealing with lots of conflicts later.
To do this, just hide the view `packages_createNewContainer` if the `FeatureFlag.SHIPPING_LABELS_M4` is not enabled, then update the PR to target `develop`.

hafizrahman committed:
Hide feature behind feature flag.

hafizrahman commented:
@hichamboushaba Thanks for the suggestion! This is now under a feature flag. Ready for another 👀

hichamboushaba reviewed (approved):
Great job @hafizrahman :shipit:
